meeting-notes/2022-05-26.md

```diff
@@ -0,0 +1,115 @@
+    - Anecdotally, this is pretty snappy!
+- A fork of the Cursorless repository (https://github.com/phillco/cursorless)
+  writes out the hats to a file (`~/.vscode-hats.json`) whenever they change.
+- Code inside of the new JetBrains plugin renders the hats inside of the editor:
+
+<img width="1397" alt="Screen Shot 2022-05-26 at 10 46 01 AM" src="https://user-images.githubusercontent.com/536668/170545671-978c35ab-d1a0-49fc-954a-89673b49008f.png">
+
+## What's not working?
+
+- VS Code only looks at the saved version of the file, so as soon as you make
+  changes, not only do they not get hats but the hats live in the wrong
+  positions.
+    - This is especially true if you add or remove lines, since they will be
+      hanging out in empty space.
+    - Unlike with VS Code, the existing hat positions don't get "rebased" as you
+      change line positions around (e.g. add a new line just before a line with
+      hats on it -- VS Code atomically moves the hats with the line; JetBrains
+      doesn't).
+- We verify that the file name of the hats file matches the file name of the
+  current editor, but otherwise it's possible to have an entirely stale hats
+  file with old data.
+- JetBrains only redraws the graphics context when it thinks it needs to (e. g.,
+  after the user types, scrolls, etc.). However, the final update of the hats in
+  reaction to that change will come later (~100ms). But JetBrains won't
+  automatically redraw the editor because it doesn't know it needs to.
+- Therefore, we have a file watcher to do this -- it watches the hats file and
+  triggers a redraw whenever it changes. But this seems to be slightly glitchy;
+  I don't think it's removing the old hats properly.
+    - You can tell because scrolling the editor slightly (which will trigger
+      JetBrains to redraw everything, but not enough to actually change the
+      hats) "fixes" things.
+- There's no way to run commands yet.
+
+## What's next for the prototype?
+
+See MILESTONES.md for most of it.
+
+## Notes / learnings:
+
+- VS Code plugins are single threaded (VS Code itself has multiple JS engines.
+  The editor has its own, but all plugins share a single one)
+- Extension interoperation: extensions can return stuff from `activate()` which
+  could be called by other extensions. See `getExtensionApi.ts`
+    - This is pretty handy, because we can have the sidecar extension (better
+      name needed) and Cursorless talked to each other as needed.
+- Every VS Code window is its own instance (with its own command server). It can
+  be difficult to figure out which window is focused/frontmost.
+    - If they ran HTTP servers, each one would need its own port.
+    - Again, probably not a problem for the Cursorless Everywhere user since
+      they won't be using VS Code otherwise.
+        - Although if we will have mixed users, we could consider running a
+          special version of VS Code just for the sidecar.
```

Comment:
Ah cool this addresses my point above re keypresses to determine active window

Reply:
Think this is probably something by want to do anyway, just to be able to control the state better. For example, we would lock down which extensions you install.

Reply:
Fwiw you can launch VSCode with a sandboxed extensions and settings dir